### DISCUSSION TITLE
Order of execution about auth_handler

### DISCUSSION THREAD

**Discussion by Chida82** [cheshire-cat-ai/core#956] [Ideas]
At present, the Auth handler always runs the core/standard version first and the custom version afterwards.
This is the case for the auth_handlers list, which is fixed.
![image](https://github.com/user-attachments/assets/2f23f030-9ce1-4f61-af88-27cb43bab6cb)

I propose to make the order dynamic.

At the moment, it is possible to use hook factory_allowed_auth_handlers which allows hanging to another list. Now we could not only hang but put before or reorder this list.  
![image](https://github.com/user-attachments/assets/45bc578a-e7a1-47c3-9674-419acd3880f5)
![image](https://github.com/user-attachments/assets/0e900071-919b-42e8-b922-73ead9d5694b)

My proposal is to apply before or after Core Auth depending on whether the list_auth_handler_default contains the CoreOnlyAuthConfig element first or not.
In some cases it may be more efficient to run the custom version first.
What do you think, can I develop it, would you accept this implementation? I understand I am not the only one who did not expect different behaviour for the 2 lists
ps: it would also be a step to consider disabling the custom version in the future, but that is not the point of the discussion at the moment!

**Comment by lucagobbi** [comment]
This proposal touches some discussions we’ve had regarding similar issues with the new auth system

Atm, when writing a custom auth handler, you need to lock your instance with two API keys (for http and ws) and a secret for JWT. While this setup may be somewhat opinionated, it’s a necessary part of the current system. This could be clarified better in the documentation

Regarding the execution order of the auth handlers, I understand why you'd want more flexibility. However, in my view, the order is usually less relevant because you are unlikely to use the same keys or tokens for different auth handlers. When a custom handler is implemented, the core handler typically fails since it lacks the necessary credentials, allowing the custom one to take over

More interesting, imho, is the possibility of overriding the entire list of allowed auth handlers. This would allow users to specify only custom handlers. Be aware that this would also exclude the admin panel, which relies on the default core auth handler for authentication. But you only have to worry about your own auth and don't have to manage the internal keys

**Reply by Chida82** [reply]
also in my opinion it should tend to be possible to have only one way of authentication and in the case of the custom, obviously knowing that I am going to lose admin and other tools, but if I dispense as a service, in my case I don't need the admin part, which I use in development or attack on the data, but I don't expose. the part exposed as a service I like more performance and lighter. 
In any case my proposal is to take a step, more in terms of performance, by changing the sorting option, because if I have an authentic custom it is because that is what I will use in 99% of calls

Translated with DeepL.com (free version)

**Comment by pieroit** [comment]
> I propose to make the order dynamic.

@Chida82  let me clarify

The list `factory_allowed_auth_handlers` is the list from which an admin choses the `custom_auth_handler`, only one, from the admin or via endpoint; that custom AuthHandler is run on the side of the `core_auth_handler`. So there will always be:
- a `core_auth_handler`, instance of [`CoreAuthHandler`](https://github.com/cheshire-cat-ai/core/blob/f14f2c0acebddd3f1f1efb6ba3b161bea45e9e3a/core/cat/factory/custom_auth_handler.py#L62) (you cannot change it or disable it from the hook)
- an additional `custom_auth_handler`, chosen from the list in `factory_allowed_auth_handlers`. If you chose no custom auth, `CoreOnlyAuthHandler` just blocks everything. If you add via plugin and setup a custom handler, for example a simple JWT validator, that auth handler will run alongside the core one above. A great example is the [Catcloak plugin](https://github.com/lucagobbi/catcloak) by @lucagobbi to have Keycloak as identity provider.

So to recap, there are 2 handlers: core, always present and a custom one you can choose form a list.

The `CoreOnlyAuthConfig` ([mapping to `CoreOnlyAuthHandler`](https://github.com/cheshire-cat-ai/core/blob/f14f2c0acebddd3f1f1efb6ba3b161bea45e9e3a/core/cat/factory/auth_handler.py#L28)) is there just as a placeholder in case no custom auth is chosen; in fact [it does not allow any request](https://github.com/cheshire-cat-ai/core/blob/f14f2c0acebddd3f1f1efb6ba3b161bea45e9e3a/core/cat/factory/custom_auth_handler.py#L157).

> My proposal is to apply before or after Core Auth depending on whether the list_auth_handler_default contains the CoreOnlyAuthConfig element first or not. In some cases it may be more efficient to run the custom version first. What do you think, can I develop it, would you accept this implementation?

Having the `custom_auth_handler` run before `core_auth_handler` can be done, agree it is more efficient in most production setups. I hope it is clear that this has nothing to do with the hook, because the hook only allows to **list** the available custom auth handlers.
You can make a PR about this, just change the order so the custom auth goes first. Please use the `develop` branch.

> I understand I am not the only one who did not expect different behaviour for the 2 lists ps: it would also be a step to consider disabling the custom version in the future, but that is not the point of the discussion at the moment!

There have been lots of discussions about this, and the most debated point was to have a custom auth without sacrificing the admin and a basic auth configuration. So now there is a core auth and a custom auth, side by side. I wanted to have a simpler hook just to list and run all auth handlers, Django style, but community decided it was "dangerous".

Let me know if everything is clear and thanks @lucagobbi for helping the clarification

P.S.: As Luca says, core auth [must be secured with env variables](https://cheshire-cat-ai.github.io/docs/production/auth/authentication/) in any case, because it will run in parallel with your custom auth
P.P.S.: hopefully in V2 we can simplify this system, it's the best so far

![image](https://github.com/user-attachments/assets/a27a6d26-7b0a-4af2-a8e2-34c72f85dd9a)

**Reply by Chida82** [reply]
Thank you for the clarification, it took me several hours to understand these concepts, they were not natural to me and even less so having to configure the core ocn keys not to use it. I recommend putting it into docuemnation in a clear and even redundant way :) .

I can do PR, but I wanted to understand first what would be appreciated. Since the two lists are not closely related, I would say not to follow my first suggestion. However I think it is useful to have a setting to drive the approach, in development it is useful to have the core first, also because you could make big mistakes in custom auth development and you would block admin portal. What do you think of an enviroment key? Something like ccat_Auth_custom_first where by default it is false

**Reply by pieroit** [reply]
> Thank you for the clarification, it took me several hours to understand these concepts, they were not natural to me and even less so having to configure the core ocn keys not to use it. I recommend putting it into docuemnation in a clear and even redundant way :) .

Yes can be greatly simplified, let's see for V2. Documentation has several warnings about this

> 
> I can do PR, but I wanted to understand first what would be appreciated. Since the two lists are not closely related, I would say not to follow my first suggestion. However I think it is useful to have a setting to drive the approach, in development it is useful to have the core first, also because you could make big mistakes in custom auth development and you would block admin portal. What do you think of an enviroment key? Something like ccat_Auth_custom_first where by default it is false

If the first handler (the custom one) blocks everything, still the core one will allow access.
I agree on having custom auth handler go first, another env variable seems overkill

**Comment by pieroit** [comment]
@Chida82 attention in `develop` and also coming soon to `main` we have the AuthHandler's methods all sync (see #973 )
Since we run all calls in threadpool, the hassle was not worth it

TLDR: change your auth handlers so they are sync
